### PR TITLE
autodiffxd tests: Establish heap use baselines

### PR DESCRIPTION
**Pull request by rpoyner-tri**
Relevant to: #13985

Extend and use LimitMalloc to measure and enforce heap use of AutoDiffXd
math operators. These limits will help clarify the effect of subsequent
changes on the implementation of AutoDiffXd.

Summary of changes

* LimitMalloc:
  * Add params() accessor and .min_num_allocations parameter
  * Adjust error messages for clarity
  * Extend unit tests to cover new features
* autodifxd_*_test.cc:
  * Add LimitMalloc guards to enforce exact heap usage

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14018)

### CODE REVIEW

**Review by rpoyner-tri — COMMENTED** [review]
This is a fun one. It may look pointless, but it sets the stage to measure the benefits of subsequent changes.
+@jwnimmer-tri for feature review, since limit-malloc changes are prominent.

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14018#-:-MGLGTd7Al_8R11lsWyE:blzxmw4)* status: LGTM missing from assignee jwnimmer-tri(platform), needs at least two assigned reviewers (waiting on @jwnimmer-tri)

**Comment by jwnimmer-tri** [comment]
___
*[common/test_utilities/limit_malloc.cc, line 201 at r1](https://reviewable.io/reviews/robotlocomotion/drake/14018#-MGMHU3-3XNU5C8TJaac:-MGMHU3-3XNU5C8TJaad:b-3o812d) ([raw file](https://github.com/robotlocomotion/drake/blob/b23c7f013a1de612c2f8c7398acf2507a1735162/common/test_utilities/limit_malloc.cc#L201)):*
> ```cpp
>   }
> 
>   if ((args.min_num_allocations >= 0) &&
> ```

In the same way that the user-provided upper bound is inert when on macOS, shouldn't the lower bound also be inert?

```c++
/// @note This class is currently a no-op on macOS.
```

That would also let us keep the BUILD file's test regexes unchanged, I think.

**Review by jwnimmer-tri — COMMENTED** [review]
CI dislikes the proposed lower bounds.

> Reviewed 4 of 27 files at r1.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14018#-:-MGMKEyQ4KelKQMyu2fF:b-c75kms)* status: 4 unresolved discussions, LGTM missing from assignee jwnimmer-tri(platform), needs at least two assigned reviewers (waiting on @jwnimmer-tri and @rpoyner-tri)

___
*[common/test_utilities/limit_malloc.cc, line 192 at r1](https://reviewable.io/reviews/robotlocomotion/drake/14018#-MGMIQqfCnicXCNOtEp7:-MGMIQqfCnicXCNOtEp8:b-lyo5e) ([raw file](https://github.com/robotlocomotion/drake/blob/b23c7f013a1de612c2f8c7398acf2507a1735162/common/test_utilities/limit_malloc.cc#L192)):*
> ```cpp
> LimitMalloc::~LimitMalloc() {
>   const int observed = num_allocations();
>   const auto args = params();
> ```

This is super tricky.  The return type is a const reference to storage that is about to evaporate on the next line.  I think it's important to make the copying explicit, not hidden in `auto`.  Probably also a comment "// Copy out the monitor's data before we delete it."
___
*[common/test_utilities/test/limit_malloc_test.cc, line 59 at r1](https://reviewable.io/reviews/robotlocomotion/drake/14018#-MGMJ84aCwrUSkO3LhLA:-MGMJ84aCwrUSkO3LhLB:bda3z37) ([raw file](https://github.com/robotlocomotion/drake/blob/b23c7f013a1de612c2f8c7398acf2507a1735162/common/test_utilities/test/limit_malloc_test.cc#L59)):*
> ```cpp
> // because exceptions cannot propagate through malloc() because it is a C
> // ABI. Similarly, the implementation of min limits must use abort() because it
> // sense failure within a destructor, which is a C++ `noexcept` context.
> ```

nit Typo `senses`.
___
*[common/test_utilities/test/limit_malloc_test.cc, line 64 at r1](https://reviewable.io/reviews/robotlocomotion/drake/14018#-MGMJbJYB_0hrkjWJ77_:-MGMJbJYB_0hrkjWJ77a:bw9ebj0) ([raw file](https://github.com/robotlocomotion/drake/blob/b23c7f013a1de612c2f8c7398acf2507a1735162/common/test_utilities/test/limit_malloc_test.cc#L64)):*
> ```cpp
> TEST_P(LimitMallocTest, UnlimitedTest) {
>   LimitMalloc guard(LimitMallocParams{ /* no limits specified */ });
>   // EXPECT_EQ() inside the guard also allocates.
> ```

nit Instead of this comment, perhaps we should be uniform within this entire file and never EXPECT with an armed guard present.  We could save the num_allocations and params and check them after disabling the guard, as we do elsewhere.

**Review by rpoyner-tri — COMMENTED** [review]
Exciting. From first sniff it looks like clang does better than gcc at avoiding the heap.

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14018#-:-MGOWQJtAMC5iokHfEJ3:bapw367)* status: 4 unresolved discussions, LGTM missing from assignee jwnimmer-tri(platform), needs at least two assigned reviewers (waiting on @jwnimmer-tri and @rpoyner-tri)

**Comment by rpoyner-tri** [comment]
*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/14018#-MGOt6a90q0Ln-d1plSo:-MGOt6a90q0Ln-d1plSp:bo03iaj) (no related file):*
@drake-jenkins-bot mac-catalina-clang-bazel-experimental-release please.
It's got the whole long weekend to finish, right? ;)

---

**Review by rpoyner-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14018)* status: 3 unresolved discussions, LGTM missing from assignee jwnimmer-tri(platform), needs at least two assigned reviewers (waiting on @jwnimmer-tri and @rpoyner-tri)

___
*[common/test_utilities/limit_malloc.cc, line 192 at r1](https://reviewable.io/reviews/robotlocomotion/drake/14018#-MGMIQqfCnicXCNOtEp7:-MGOsfz5-bXwZDCUtLhj:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/b23c7f013a1de612c2f8c7398acf2507a1735162/common/test_utilities/limit_malloc.cc#L192)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

This is super tricky.  The return type is a const reference to storage that is about to evaporate on the next line.  I think it's important to make the copying explicit, not hidden in `auto`.  Probably also a comment "// Copy out the monitor's data before we delete it."
</blockquote></details>

Done.
___
*[common/test_utilities/limit_malloc.cc, line 201 at r1](https://reviewable.io/reviews/robotlocomotion/drake/14018#-MGMHU3-3XNU5C8TJaac:-MGOsi970CmYB0YE-6lT:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/b23c7f013a1de612c2f8c7398acf2507a1735162/common/test_utilities/limit_malloc.cc#L201)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

In the same way that the user-provided upper bound is inert when on macOS, shouldn't the lower bound also be inert?

```c++
/// @note This class is currently a no-op on macOS.
```

That would also let us keep the BUILD file's test regexes unchanged, I think.
</blockquote></details>

Done.

**Review by jwnimmer-tri — COMMENTED** [review]
The test_utilities changes are looking good now.

The CI calibration still needs more work.

> Reviewed 2 of 5 files at r2.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14018#-:-MGOxXiB5KLTsl1ZnSaT:b81glu0)* status: LGTM missing from assignee jwnimmer-tri(platform), needs at least two assigned reviewers (waiting on @jwnimmer-tri)

**Comment by rpoyner-tri** [comment]
*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/14018#-MGOt6a90q0Ln-d1plSo:-MGQ06v27OaKSx8P1UAd:b-e2oap8) (no related file):*
<details><summary><i>Previously, rpoyner-tri (Rick Poyner (rico)) wrote…</i></summary><blockquote>

@drake-jenkins-bot mac-catalina-clang-bazel-experimental-release please.
It's got the whole long weekend to finish, right? ;)
</blockquote></details>

@drake-jenkins-bot mac-catalina-clang-bazel-experimental-release please.

---

**Comment by rpoyner-tri** [comment]
*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/14018#-MGOt6a90q0Ln-d1plSo:-MGTypnd5eftwLQvZKqQ:b-e2oap8) (no related file):*
<details><summary><i>Previously, rpoyner-tri (Rick Poyner (rico)) wrote…</i></summary><blockquote>

@drake-jenkins-bot mac-catalina-clang-bazel-experimental-release please.
</blockquote></details>

@drake-jenkins-bot mac-catalina-clang-bazel-experimental-release please.

---

**Review by rpoyner-tri — COMMENTED** [review]
The next two scheduled platform reviewers are already implicated. Tossing it out to +@sherm1 since he's probably already read it.

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14018#-:-MGU2FB2CaOUgxa-5aPs:b-fv5ol2)* status: LGTM missing from assignees jwnimmer-tri(platform),sherm1(platform) (waiting on @jwnimmer-tri and @sherm1)

**Review by sherm1 — COMMENTED** [review]
I'm curious how you plan to make use of the "min allocations"? Is that so you'll be able to tell if you've made improvemements that you might not otherwise have noticed?

Anyway, looks great: platform :lgtm:.

> Reviewed 21 of 27 files at r1, 2 of 5 files at r2, 3 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14018#-:-MGVhl3SCHbvh-hIhXXG:b24nlg4)* status: LGTM missing from assignee jwnimmer-tri(platform)

**Review by rpoyner-tri — COMMENTED** [review]
Yes, the min violation message is a great convenience for finding the actual value.

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14018#-:-MGZhQJj9CiwUOZzX8jR:b-ppno0o)* status: LGTM missing from assignee jwnimmer-tri(platform)

**Review by jwnimmer-tri — COMMENTED** [review]
> Reviewed 20 of 27 files at r1, 2 of 5 files at r2, 2 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14018)* status: 1 unresolved discussion, LGTM missing from assignee jwnimmer-tri(platform) (waiting on @rpoyner-tri)

___
*[common/test/autodiffxd_abs2_test.cc, line 9 at r3](https://reviewable.io/reviews/robotlocomotion/drake/14018#-MGacK__A8Zrk5dSwh2l:-MGacK__A8Zrk5dSwh2m:bpom05k) ([raw file](https://github.com/robotlocomotion/drake/blob/20f9eff59c4aaaf5672c620ac36812d680aa90ec/common/test/autodiffxd_abs2_test.cc#L9)):*
> ```cpp
> namespace {
> 
> TEST_F(AutoDiffXdTest, Abs2) {
> ```

_Meta_

The pre-existing test cases are focused on logical correctness and functional equivalence.  I could easily see us adding or changing the `CHECK_UNARY_FUNCTION` implementation to add or remove test cases, or that a change to comparison reporting failure (how `operator<<` for test case annotations works, or how `CompareMatrices` is implemented) could disturb these results without any actual changes in the `common/autodiff.h` implementation.

In short, it seems to me like adding heap limits to correctness tests is too brittle.

My suggestion is to add a dedicated `autodiffxd_heap_test` that covers precisely what you need.

**Review by rpoyner-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14018)* status: 1 unresolved discussion, LGTM missing from assignee jwnimmer-tri(platform) (waiting on @jwnimmer-tri and @sherm1)

___
*[common/test/autodiffxd_abs2_test.cc, line 9 at r3](https://reviewable.io/reviews/robotlocomotion/drake/14018#-MGacK__A8Zrk5dSwh2l:-MGnak9Z2l9od3N-y5kR:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/20f9eff59c4aaaf5672c620ac36812d680aa90ec/common/test/autodiffxd_abs2_test.cc#L9)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

_Meta_

The pre-existing test cases are focused on logical correctness and functional equivalence.  I could easily see us adding or changing the `CHECK_UNARY_FUNCTION` implementation to add or remove test cases, or that a change to comparison reporting failure (how `operator<<` for test case annotations works, or how `CompareMatrices` is implemented) could disturb these results without any actual changes in the `common/autodiff.h` implementation.

In short, it seems to me like adding heap limits to correctness tests is too brittle.

My suggestion is to add a dedicated `autodiffxd_heap_test` that covers precisely what you need.
</blockquote></details>

Done.

**Review by jwnimmer-tri — COMMENTED** [review]
FYI `bazel test --config lint //...` is a great shortcut to have at hand.

> Reviewed 1 of 3 files at r3, 3 of 24 files at r4.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14018#-:-MGncJCs2qhLlvSH812Z:b4v6cnb)* status: 2 unresolved discussions, LGTM missing from assignee jwnimmer-tri(platform) (waiting on @rpoyner-tri)

___
*[common/test/autodiffxd_heap_test.cc, line 28 at r4](https://reviewable.io/reviews/robotlocomotion/drake/14018#-MGnci5qEIUP4NKY3aLS:-MGnci5qEIUP4NKY3aLT:baeeqnl) ([raw file](https://github.com/robotlocomotion/drake/blob/a4a6f64d6eb61a8571bafecbca15d3313ec2e081/common/test/autodiffxd_heap_test.cc#L28)):*
> ```cpp
> TEST_F(AutoDiffXdHeapTest, Abs) {
>   LimitMalloc guard({.max_num_allocations = 2, .min_num_allocations = 2});
>   volatile auto v = abs(x_ + y_);
> ```

It is not clear to me what purpose `volatile auto` here is serving.  If we wrote `abs(x_ + y_);` as the statement without any assignment, should I infer that the compiler could choose to dead-code-eliminate the entire expression?  Or is it an rvalue vs lvalue thing?  This probably needs another note comment above, to explain for future readers.
___
*[common/test_utilities/limit_malloc.cc, line 164 at r4](https://reviewable.io/reviews/robotlocomotion/drake/14018#-MGnhj-y7dOiU8NyePDg:-MGnhj-y7dOiU8NyePDh:b-g4gx60) ([raw file](https://github.com/robotlocomotion/drake/blob/a4a6f64d6eb61a8571bafecbca15d3313ec2e081/common/test_utilities/limit_malloc.cc#L164)):*
> ```cpp
> 
>   if (!failure) { return; }
>   if (IsLeakSanitizerBuild()) { return; }
> ```

Hmm.  If our limit enforcement is disarmed because it lsan turns it into lies, why are we still incrementing `observed_num_allocations_` above?  Shouldn't this bail-out check happen as the first thing in the method?

**Review by rpoyner-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14018)* status: 2 unresolved discussions, LGTM missing from assignee jwnimmer-tri(platform) (waiting on @jwnimmer-tri and @rpoyner-tri)

___
*[common/test_utilities/limit_malloc.cc, line 164 at r4](https://reviewable.io/reviews/robotlocomotion/drake/14018#-MGnhj-y7dOiU8NyePDg:-MGnkw_g7J0ZtYTYIwaO:b-fw8qa) ([raw file](https://github.com/robotlocomotion/drake/blob/a4a6f64d6eb61a8571bafecbca15d3313ec2e081/common/test_utilities/limit_malloc.cc#L164)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Hmm.  If our limit enforcement is disarmed because it lsan turns it into lies, why are we still incrementing `observed_num_allocations_` above?  Shouldn't this bail-out check happen as the first thing in the method?
</blockquote></details>

My motivation was to keep that check out of the fast path. But, as you point out, the num_allocations() can still observe something happening, even if it is wrong. Will look again.

**Review by jwnimmer-tri — COMMENTED** [review]
> Reviewed 2 of 2 files at r5.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14018)* status: 2 unresolved discussions, LGTM missing from assignee jwnimmer-tri(platform) (waiting on @rpoyner-tri)

___
*[common/test_utilities/limit_malloc.cc, line 164 at r4](https://reviewable.io/reviews/robotlocomotion/drake/14018#-MGnhj-y7dOiU8NyePDg:-MGnlza6DtXsq9xBAdzn:b-7xhgvz) ([raw file](https://github.com/robotlocomotion/drake/blob/a4a6f64d6eb61a8571bafecbca15d3313ec2e081/common/test_utilities/limit_malloc.cc#L164)):*
<details><summary><i>Previously, rpoyner-tri (Rick Poyner (rico)) wrote…</i></summary><blockquote>

My motivation was to keep that check out of the fast path. But, as you point out, the num_allocations() can still observe something happening, even if it is wrong. Will look again.
</blockquote></details>

Ah.  It would be fine to memoize (and even pre-warm) it's return value using `never_destroyed`, so that we're not calling back into the stdlib during a malloc.

**Review by rpoyner-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14018)* status: 2 unresolved discussions, LGTM missing from assignee jwnimmer-tri(platform) (waiting on @rpoyner-tri)

___
*[common/test_utilities/limit_malloc.cc, line 164 at r4](https://reviewable.io/reviews/robotlocomotion/drake/14018#-MGnhj-y7dOiU8NyePDg:-MGnmV3l6l0FafsrwVXG:bdpvf26) ([raw file](https://github.com/robotlocomotion/drake/blob/a4a6f64d6eb61a8571bafecbca15d3313ec2e081/common/test_utilities/limit_malloc.cc#L164)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Ah.  It would be fine to memoize (and even pre-warm) it's return value using `never_destroyed`, so that we're not calling back into the stdlib during a malloc.
</blockquote></details>

Is never-destroyed necessary for a function-scoped static bool?

**Review by rpoyner-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14018)* status: 2 unresolved discussions, LGTM missing from assignee jwnimmer-tri(platform) (waiting on @rpoyner-tri)

___
*[common/test_utilities/limit_malloc.cc, line 164 at r4](https://reviewable.io/reviews/robotlocomotion/drake/14018#-MGnhj-y7dOiU8NyePDg:-MGnn_IZ7N3BR8q5b7RB:b-6w8q5b) ([raw file](https://github.com/robotlocomotion/drake/blob/a4a6f64d6eb61a8571bafecbca15d3313ec2e081/common/test_utilities/limit_malloc.cc#L164)):*
<details><summary><i>Previously, rpoyner-tri (Rick Poyner (rico)) wrote…</i></summary><blockquote>

Is never-destroyed necessary for a function-scoped static bool?
</blockquote></details>

Ah. pre-warm.

**Review by jwnimmer-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14018)* status: 2 unresolved discussions, LGTM missing from assignee jwnimmer-tri(platform) (waiting on @rpoyner-tri)

___
*[common/test_utilities/limit_malloc.cc, line 164 at r4](https://reviewable.io/reviews/robotlocomotion/drake/14018#-MGnhj-y7dOiU8NyePDg:-MGnq1OhC6F9PAnWwPJf:bvgybg) ([raw file](https://github.com/robotlocomotion/drake/blob/a4a6f64d6eb61a8571bafecbca15d3313ec2e081/common/test_utilities/limit_malloc.cc#L164)):*
<details><summary><i>Previously, rpoyner-tri (Rick Poyner (rico)) wrote…</i></summary><blockquote>

Ah. pre-warm.
</blockquote></details>

I'm not tied to `never_destroyed` in particular.  Any clear and correct code to memoize is fine.

**Review by rpoyner-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14018)* status: 2 unresolved discussions, LGTM missing from assignee jwnimmer-tri(platform) (waiting on @jwnimmer-tri, @rpoyner-tri, and @sherm1)

___
*[common/test_utilities/limit_malloc.cc, line 164 at r4](https://reviewable.io/reviews/robotlocomotion/drake/14018#-MGnhj-y7dOiU8NyePDg:-MGnroVlE-iIdUamAHwD:b1lre7r) ([raw file](https://github.com/robotlocomotion/drake/blob/a4a6f64d6eb61a8571bafecbca15d3313ec2e081/common/test_utilities/limit_malloc.cc#L164)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I'm not tied to `never_destroyed` in particular.  Any clear and correct code to memoize is fine.
</blockquote></details>

Updated. PTAL.

**Review by rpoyner-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14018)* status: 2 unresolved discussions, LGTM missing from assignee jwnimmer-tri(platform) (waiting on @jwnimmer-tri, @rpoyner-tri, and @sherm1)

___
*[common/test/autodiffxd_heap_test.cc, line 28 at r4](https://reviewable.io/reviews/robotlocomotion/drake/14018#-MGnci5qEIUP4NKY3aLS:-MGntEidCxs2Er_gl7YR:b84m43l) ([raw file](https://github.com/robotlocomotion/drake/blob/a4a6f64d6eb61a8571bafecbca15d3313ec2e081/common/test/autodiffxd_heap_test.cc#L28)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

It is not clear to me what purpose `volatile auto` here is serving.  If we wrote `abs(x_ + y_);` as the statement without any assignment, should I infer that the compiler could choose to dead-code-eliminate the entire expression?  Or is it an rvalue vs lvalue thing?  This probably needs another note comment above, to explain for future readers.
</blockquote></details>

All of the tested functions are marked inline. While i don't have a working test case where the optimizer eats the function, neither do I have any good assurances that it won't. I'll write a note.

**Review by jwnimmer-tri — COMMENTED** [review]
:lgtm: feature.

> Reviewed 1 of 1 files at r6.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14018#-:-MGnuQTxCXOJoHTohs1j:b-n37teh)* status: 2 unresolved discussions (waiting on @rpoyner-tri)

*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/14018#-MGnuXakE2FUkULxk0I-:-MGnuXakE2FUkULxk0I0:bnc4h8m) (no related file):*
nit This needs another look from @sherm1, given the rewrites.

---
___
*[common/test/autodiffxd_heap_test.cc, line 28 at r4](https://reviewable.io/reviews/robotlocomotion/drake/14018#-MGnci5qEIUP4NKY3aLS:-MGntzPI9kee9MuAyEq_:b-ucfkn4) ([raw file](https://github.com/robotlocomotion/drake/blob/a4a6f64d6eb61a8571bafecbca15d3313ec2e081/common/test/autodiffxd_heap_test.cc#L28)):*
<details><summary><i>Previously, rpoyner-tri (Rick Poyner (rico)) wrote…</i></summary><blockquote>

All of the tested functions are marked inline. While i don't have a working test case where the optimizer eats the function, neither do I have any good assurances that it won't. I'll write a note.
</blockquote></details>

Thanks.

**Review by rpoyner-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14018)* status: 2 unresolved discussions (waiting on @jwnimmer-tri and @sherm1)

*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/14018#-MGnuXakE2FUkULxk0I-:-MGnwXB_3QwCj0_Oo163:bi6n0b5) (no related file):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

nit This needs another look from @sherm1, given the rewrites.
</blockquote></details>

agreed.

---
___
*[common/test/autodiffxd_heap_test.cc, line 28 at r4](https://reviewable.io/reviews/robotlocomotion/drake/14018#-MGnci5qEIUP4NKY3aLS:-MGnwOnqDaD9ODMc0k65:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a4a6f64d6eb61a8571bafecbca15d3313ec2e081/common/test/autodiffxd_heap_test.cc#L28)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Thanks.
</blockquote></details>

Done.

**Review by jwnimmer-tri — COMMENTED** [review]
> Reviewed 1 of 1 files at r7.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14018)* status: 1 unresolved discussion (waiting on @sherm1)

___
*[common/test/autodiffxd_heap_test.cc, line 28 at r4](https://reviewable.io/reviews/robotlocomotion/drake/14018#-MGnci5qEIUP4NKY3aLS:-MGnxpcX9fryWoR4dJjG:bcp03c1) ([raw file](https://github.com/robotlocomotion/drake/blob/a4a6f64d6eb61a8571bafecbca15d3313ec2e081/common/test/autodiffxd_heap_test.cc#L28)):*
<details><summary><i>Previously, rpoyner-tri (Rick Poyner (rico)) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

AIUI the `inline` keyword is not relevant -- it's about ODR, not optimization.

**Review by sherm1 — COMMENTED** [review]
Platform re-:lgtm:

> Reviewed 1 of 2 files at r5, 1 of 1 files at r6, 1 of 1 files at r8.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14018#-:-MGo4sxyBavb2jX_voXp:b8vw1y6)* status: 2 unresolved discussions (waiting on @rpoyner-tri and @sherm1)

___
*[common/test_utilities/limit_malloc.cc, line 26 at r8](https://reviewable.io/reviews/robotlocomotion/drake/14018#-MGo3JYG-MfzPW3GGeuu:-MGo3JYG-MfzPW3GGeuv:bxr6o6z) ([raw file](https://github.com/robotlocomotion/drake/blob/4acf54b08af28df0504ff9f2288eb1afe4f8f937/common/test_utilities/limit_malloc.cc#L26)):*
> ```cpp
> // LimitMalloc does not work properly with leak sanitizer builds. Check this
> // predicate and disarm to avoid erroneous results.
> static bool IsLeakSanitizerBuild() {
> ```

BTW why have these two functions as :: scope statics rather than inside the anonymous namespace below?

**Review by rpoyner-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14018)* status: 2 unresolved discussions (waiting on @sherm1)

___
*[common/test_utilities/limit_malloc.cc, line 26 at r8](https://reviewable.io/reviews/robotlocomotion/drake/14018#-MGo3JYG-MfzPW3GGeuu:-MGo5V446E4Smyd52eUw:b-4kykya) ([raw file](https://github.com/robotlocomotion/drake/blob/4acf54b08af28df0504ff9f2288eb1afe4f8f937/common/test_utilities/limit_malloc.cc#L26)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW why have these two functions as :: scope statics rather than inside the anonymous namespace below?
</blockquote></details>

They are called by other global statics, for the usual cascade of annoying reasons. EvaluateMinNumAllocations() gets no-oped at compile time by participating in the `#fidef APPLE` stanza. All of that stuff has to be global for the C link seam to `malloc` to work.  isLeakSanitizerBuild() gets checked within EvalulateminNumAllocations(). I suppose it could be moved into the anon namespace, but I'm not feeling super strongly about it.
